### PR TITLE
Add default bearer signer and stubs for other signers

### PR DESCRIPTION
**Pull request by haydenbaker**
## Motivation and Context
- See below

## Modifications
- Add a default implementation of the `BearerHttpSigner` and tests
- Add stubs for all signer default implementations
- Fix up signer interface definitions to include identity (needed after update to HttpSigner interface)
- Fix up POMs to address some small nits

## Screenshots (if appropriate)
N/A

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## Checklist

- [X] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [X] Local run of `mvn install` succeeds
- [X] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [X] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [X] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [X] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/http-auth-aws/pom.xml [hunk 1]
```diff
@@ -28,8 +28,8 @@
     <artifactId>http-auth-aws</artifactId>
     <name>AWS Java SDK :: HTTP Auth AWS</name>
     <description>
-        The AWS SDK for Java - HTTP Auth AWS module contains interfaces and implementations for http authentication
-            specific to AWS.
+        The AWS SDK for Java - HTTP Auth AWS module contains interfaces and implementations for generic HTTP
```
Comment (gosar) [inline core/http-auth-aws/pom.xml:31]: drop generic here

File: core/http-auth-event-stream/pom.xml [hunk 2]
```diff
@@ -28,7 +28,7 @@
     <artifactId>http-auth-event-stream</artifactId>
     <name>AWS Java SDK :: HTTP Auth Event Stream</name>
     <description>
-        The AWS SDK for Java - HTTP Auth module contains interfaces and implementations for http authentication
+        The AWS SDK for Java - HTTP Auth module contains interfaces and implementations for generic HTTP authentication
```
Comment (gosar) [inline core/http-auth-event-stream/pom.xml:31]: Maybe:
The AWS SDK for Java - HTTP Auth Event Stream module contains interfaces and implementations for authentication of event streams in HTTP services.

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/BearerHttpSigner.java [hunk 3]
```diff
@@ -16,9 +16,18 @@
...
 
+/**
+ * An {@link HttpSigner} that will sign a request with a
+ * bearer-token ({@link TokenIdentity}).
+ */
 @SdkPublicApi
-public interface BearerHttpSigner extends HttpSigner {
+public interface BearerHttpSigner extends HttpSigner<TokenIdentity> {
 
+    static BearerHttpSigner create() {
```
Comment (gosar) [inline core/http-auth/src/main/java/software/amazon/awssdk/http/auth/BearerHttpSigner.java:30]: add javadoc

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/AwsV4QueryHttpSigner.java [hunk 4]
```diff
@@ -17,8 +17,11 @@
...
 import software.amazon.awssdk.annotations.SdkPublicApi;
 import software.amazon.awssdk.http.auth.spi.HttpSigner;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
 
 @SdkPublicApi
-public interface AwsV4QueryHttpSigner extends HttpSigner {
-
+public interface AwsV4QueryHttpSigner extends HttpSigner<AwsCredentialsIdentity> {
+    static AwsV4QueryHttpSigner create() {
+        throw new UnsupportedOperationException();
```
Comment (gosar) [inline core/http-auth/src/main/java/software/amazon/awssdk/http/auth/AwsV4QueryHttpSigner.java:25]: DefaultAwsV4QueryHttpSigner

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                .request(doSign(request))
+                .payload(request.payload().orElse(null))
+                .build();
+    }
+
+    /**
+     * Using {@link HttpSignRequest}, sign the request with
+     * a {@link HttpSignRequest} and re-build it.
+     */
+    private <P, I extends TokenIdentity> SdkHttpRequest doSign(HttpSignRequest<P, I> request) {
```
Comment (gosar) [inline core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java:58]: nit: given P and I aren't used specifically, wonder about this v/s
`private SdkHttpRequest doSign(HttpSignRequest<?, ? extends TokenIdentity> request) {`

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/AwsV4QueryHttpSigner.java (hunk 4 above)
Comment (haydenbaker) [reply core/http-auth/src/main/java/software/amazon/awssdk/http/auth/AwsV4QueryHttpSigner.java:25]: Yep. Missed adding that in the commit.

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java (hunk 5 above)
Comment (haydenbaker) [reply core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java:58]: Sure. Done.

File: core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java [hunk 6]
```diff
@@ -16,9 +16,24 @@
 package software.amazon.awssdk.http.auth.aws.crt;
 
 import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.http.auth.aws.crt.internal.DefaultAwsCrtS3V4aHttpSigner;
 import software.amazon.awssdk.http.auth.spi.HttpSigner;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
 
+/**
+ * An {@link HttpSigner} that will sign a request
```
Comment (gosar) [inline core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java:24]: I think the style used is 130 for line length, so can reduce the number of lines for these.

File: core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java [hunk 7]
```diff
@@ -16,9 +16,24 @@
...
 @SdkPublicApi
-public interface AwsCrtV4aHttpAuthScheme extends HttpSigner {
+public interface AwsCrtS3V4aHttpSigner extends HttpSigner<AwsCredentialsIdentity> {
 
+    /**
+     * Get a default implementation of a {@link AwsCrtS3V4aHttpSigner}
+     *
+     * @return AwsCrtS3V4aHttpSigner
+     */
+    static AwsCrtS3V4aHttpSigner create() {
```
Comment (gosar) [inline core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java:36]: I'd like Java team to confirm whether these methods should be called create or defaultSigner?

File: core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java [hunk 8]
```diff
@@ -16,9 +16,24 @@
...
+ * An {@link HttpSigner} that will sign a request
+ * using an AWS credentials ({@link AwsCredentialsIdentity}),
+ * specifically for S3-CRT.
+ */
 @SdkPublicApi
-public interface AwsCrtV4aHttpAuthScheme extends HttpSigner {
+public interface AwsCrtS3V4aHttpSigner extends HttpSigner<AwsCredentialsIdentity> {
 
+    /**
+     * Get a default implementation of a {@link AwsCrtS3V4aHttpSigner}
```
Comment (gosar) [inline core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java:32]: I think the style is to end these with `.`.

File: core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java (hunk 7 above)
Comment (davidh44) [reply core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/AwsCrtS3V4aHttpSigner.java:36]: `create()` is fine

**Review by davidh44 — APPROVED** [review]

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                        AUTHZ_HEADER,
+                        buildAuthorizationHeader(request.identity()))
+                .build();
+    }
+
+    /**
+     * Use a {@link TokenIdentity} to build an authorization header.
+     */
+    private String buildAuthorizationHeader(TokenIdentity tokenIdentity) {
+        return String.format("%s %s", BEARER_LABEL, tokenIdentity.token());
```
Comment (zoewangg) [inline core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java:69]: Micro performance optimization: can we just use `BEARER_LABEL + " " + tokenIdentity.token()`?

**Review by gosar — APPROVED** [review]

File: core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java (hunk 9 above)
Comment (haydenbaker) [reply core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultBearerHttpSigner.java:69]: Sure, thanks for the suggestion. I'll keep this in mind for other signers as well.

**Review by zoewangg — APPROVED** [review]

**Pull request merged**